components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt

```diff
@@ -471,9 +516,51 @@ class SystemEngineView @JvmOverloads constructor(
         }
     }
 
+    private fun resetJSAlertAbuseState() {
+        alertJSCount = 0
+        userWantToShowMoreDialogs = true
+    }
+
+    internal fun updateJSDialogAbusedState() {
+        if (!areDialogsBeingAbusedByTime()) {
+            ++alertJSCount
+            lastTimeWhenAJSDialogWasShown = Date()
+        } else {
+            alertJSCount = 0
+        }
+    }
+
+    internal fun areDialogsBeingAbused(): Boolean {
+        return areDialogsBeingAbusedByTime() || areDialogsBeingAbusedByAmount()
+    }
+
+    @Suppress("MagicNumber")
+    internal fun areDialogsBeingAbusedByTime(): Boolean {
+        return if (alertJSCount == 0) {
+            false
+        } else {
+            val now = Date()
+            val diffInSeconds =
+                (now.time - lastTimeWhenAJSDialogWasShown.time) / 1000 // 1 second equal to 1000 milliseconds
+
+            diffInSeconds > MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT
```

Comment:
Hm interesting. Is this the right logic? We're saying if there's too much time between alerts then the site is abusing it. I would have expected this to be the inverse. If the time between alerts is really short i.e. it shows alerts in short succession then that's what we want to prevent?

nit: I also don't quite like the `abuseBy` naming. Maybe we can find alternatives e.g. `dialogsWithinTimeLimit` or `dialogsWithinLimit` etc. Not sure.

Reply:
@csadilek `diffInSeconds` tells how many seconds has passed from the actual time and the last time that dialog was shown. If the amount of seconds is bigger that limit (`MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT`) then the page is being abusive.

For instance if you have:
now = 11:59:40
lastDialog = 11:59:32

Then **diffInSeconds** will be **8**, 8 is bigger than 

`8  > MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT (3) ` then page is been abusive.

 ```kotlin
   @Test
    fun `are dialogs being abused by time`() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val engineView = SystemEngineView(context)

        with(engineView) {

            assertFalse(areDialogsBeingAbusedByTime())

            alertJSCount = 1

            // Decreasing lastTimeWhenAJSDialogWasShown by a year
            lastTimeWhenAJSDialogWasShown = lastTimeWhenAJSDialogWasShown.add(SECOND, - (MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT+1))

            assertTrue(areDialogsBeingAbusedByTime())
        }
    }
```
If the algorithm is wrong please let me know :)

Reply:
@jonalmeida   I had to change the logic a bit because they were doing something that I don't like. Changing state in a method that was supposed to give a value, in my opinion, this is misleading.

```c
TimeDuration dialogInterval(TimeStamp::Now() - mLastDialogQuitTime);
  if (dialogInterval.ToSeconds() <
      Preferences::GetInt("dom.successive_dialog_time_limit",
                          DEFAULT_SUCCESSIVE_DIALOG_TIME_LIMIT)) {
    mDialogAbuseCount++;

    return GetPopupControlState() > openAllowed ||
           mDialogAbuseCount > MAX_SUCCESSIVE_DIALOG_COUNT;
  } 
```

For this reason the operator are inverted, `areDialogsBeingAbusedByTime()` is just querying and `updateJSDialogAbusedState()` is mutating state.

Reply:
> If the amount of seconds is **bigger that limit** (MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT) then the page is being abusive.

I think you mean if the seconds are _less than_ the limit right? That's the abusive behaviour.

If there are a long delays between the two dialog requests this should be considered reasonable behaviour.

Reply:
@csadilek and @jonalmeida  you are right, I will inverter the condition, thanks for catching it up. I was taking **MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT** as a max diff between the too times.